README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# lofi <img src="man/figures/logo.png" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/dependencies-zero-blue.svg) [![Travis
build
status](https://travis-ci.org/coolbutuseless/lofi.svg?branch=master)](https://travis-ci.org/coolbutuseless/lofi)
[![Codecov test
coverage](https://codecov.io/gh/coolbutuseless/lofi/branch/master/graph/badge.svg)](https://codecov.io/gh/coolbutuseless/lofi?branch=master)
<!-- badges: end -->

The goal of `lofi` is to squeeze multiple, low-fidelity representations
of colours and numbers into the 32-bits of a single, standard integer in
R.

This low-fidelity representation of values (a.k.a. *lofi*) is usually
only an approximation of the original values, and reconstructed values
will most likely be slightly different from the original.

Visit the [webpage](https://coolbutuseless.github.io/package/lofi/) for
more in-depth documentation.

## What’s in the box

  - `pack()` and `unpack()` are the key functions for packing/unpacking
    multiple values into the bits of an integer
  - There is a suite of low-level functions for handling each particular
    supported type e.g. `dbl_to_lofi()` and `lofi_to_dbl()`

## Installation

You can install from [GitHub](https://github.com/coolbutuseless/lofi)
with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/lofi")
```

# `pack()` and `unpack()`

The `pack()` function will encode a set of values into the bits of a
single integer. The `unpack()` function will reconstruct the original
values from this single integer.

The schematic below illustrates the method by which a list of values is
packed into a single integer and then unpacked back into the list of
values. `pack()` and `unpack()` in turn rely on some lower level
functions `X_to_lofi()` and `lofi_to_X()` (as indicated in the yellow
boxes).

![](man/figures/pack-unpack.png)

The key to the process is the **packing specification** (a.k.a.
`pack_spec`) which defines the type of value you want to store how many
bits are used to store each value.

Steps for using this package:

1.  Define a `pack_spec` defining how each value is converted to lofi
2.  Call `pack()` on a named list of values, along with the `pack_spec`
3.  Receive back a single integer containing all the packed lofi bits
4.  Call `unpack()` on the single integer (along with the `pack_spec`)
5.  Receive a list of reconstructed values which are the same as the
    original values - but maybe with some loss of precision.

The `pack_spec` defines information on packing the following
types:

| type    | lossless? | nbits | description                                   | signed   |
| ------- | --------- | ----- | --------------------------------------------- | -------- |
| integer | Yes       | 1-32  | pack standard integer                         | optional |
| logical | Yes       | 1-32  | standard 1-bit logical. zero-padded if needed | NA       |
| choice  | Yes       | 1-32  | almost like a factor representation           | NA       |
| double  | No        | 1-32  | pack a standard double                        | optional |
| colour  | No        | 3-24  | pack a hex colour e.g. \#123456               | NA       |
| scaled  | No        | 1-32  | pack a range into the given bits              | NA       |
| custom  | Possibly  | 1-32  | user specified functions used to pack/unpack  | NA       |

The integer, logical and choice types are lossless, and original values
can be perfectly reconstructed by `unpack()`. The double, colour and
scaled types all quantize the inputs in some way and lose information -
thus the original value is always *imperfectly* reconstructed (except in
very particular circumstances).

For information on the specification for each type, see `?lofi::pack` or
[`vignette("packing-specification",
package='lofi'`](https://coolbutuseless.github.io/package/lofi/articles/packing-specification.html)

## Example - `pack/unpack` the first row of `iris` data

The `iris` dataset gives the measurements in cm of the variables sepal
length and width, and petal length and width, respectively, for 50
flowers from each of 3 species of iris. The first rows of the data are
shown below:

| Sepal.Length | Sepal.Width | Petal.Length | Petal.Width | Species |
| -----------: | ----------: | -----------: | ----------: | :------ |
|          5.1 |         3.5 |          1.4 |         0.2 | setosa  |
|          4.9 |         3.0 |          1.4 |         0.2 | setosa  |
|          4.7 |         3.2 |          1.3 |         0.2 | setosa  |

First rows of iris data

The `pack_spec` for the data seen in iris is:

  - `Sepal.Length` is a floating point value with 1 decimal place with a
    maximum value of 7.9. This could be multiplied by 10, converted to
    an integer and stored in 7 bits.
  - Similarly for `Sepal.Width`, `Petal.Length` and `Petal.Width` -
    after multiplying by 10, and treating as an integer, these values
    could all by stored in 6, 7, and 5 bits respectively.
  - `Species` is a choice from 3 options, so in the best case we only
    need 2 bits to store this information.

The defined `pack_spec` is stored as a
list:

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Can perfectly pack 'iris' into 27 bits per row.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pack_spec <- list(
  Sepal.Length = list(type = 'integer', nbits = 7, mult = 10, signed = FALSE),
  Sepal.Width  = list(type = 'integer', nbits = 6, mult = 10, signed = FALSE),
  Petal.Length = list(type = 'integer', nbits = 7, mult = 10, signed = FALSE),
  Petal.Width  = list(type = 'integer', nbits = 5, mult = 10, signed = FALSE),
  Species      = list(type = 'choice' , nbits = 2,
                      options = c('setosa', 'versicolor', 'virginica'))
)
```

Now take the first row of `iris` and `pack()` it:

``` r
lofi::pack(iris[1, ], pack_spec)
#> [1] 54052616
```

So the first row of iris has now been packed into the integer:
**54052616**.  
If this integer is viewed as the 32 bits which make it up, the different
lofi data representations can be identified:

![](man/figures/iris-bits.png)

If the integer is now **`unpack()ed`**, we get back the original data.

``` r
lofi::unpack(54052616L, pack_spec)
#> $Sepal.Length
#> [1] 5.1
#> 
#> $Sepal.Width
#> [1] 3.5
#> 
#> $Petal.Length
#> [1] 1.4
#> 
#> $Petal.Width
#> [1] 0.2
#> 
#> $Species
#> [1] "setosa"
```

For an example of using `lofi` to pack an entire data.frame see
[`vignette("packing-a-data-frame",
package='lofi')`](https://coolbutuseless.github.io/package/lofi/articles/packing-a-data-frame.html)

# Conversion to/from low-fidelity representation

Underneath `pack()` an `unpack()` is a suite of low-level functions for
handling each particular supported type

  - `dbl_to_lofi()`, `lofi_to_dbl()`
  - `int32_to_lofi()`, `lofi_to_int32()`
  - `hex_colour_to_lofi()`, `lofi_to_hex_colour()`
  - `lgl_to_lofi()`, `lofi_to_lgl()`
  - `choice_to_lofi()`, `lofi_to_choice()`

## 64-bit (double precision) floating point to Lofi

Double precision floating point values are converted to low-fidelity
representation by truncating the mantissa, and re-encoding the exponent.
Low-fidelity floats have limited range, poorer precision, and will
almost never give back the exact starting value when `unpack()ed`.

Note: `lofi` has no explicit support for `NA`, `NaN`, `Inf` or
[denormalized numbers](https://en.wikipedia.org/wiki/Denormal_number).

The following converts a double into a 10 bit float (with a sign bit,
2-bit exponent and 7-bit mantissa). The reconstructed double is close to
the original value, but not an exact
match.

| Representation                                                    | Bits | Value      | Bit layout                          |
| ----------------------------------------------------------------- | ---- | ---------- | ----------------------------------- |
| Double precision                                                  | 64   | \-1.234    | ![](man/figures/compact-float1.png) |
| Lofi double `dbl_to_lofi(-1.234, float_bits = c(1, 2, 7))`        | 10   | 669L       | ![](man/figures/compact-float2.png) |
| Reconstructed double `lofi_to_dbl(669L, float_bits = c(1, 2, 7))` | 64   | \-1.226562 | ![](man/figures/compact-float3.png) |

For more example of using `lofi` to pack double precision floating point
values see [`vignette("lofi-double",
package='lofi')`](https://coolbutuseless.github.io/package/lofi/articles/lofi-double.html)

## RGB Hex Colour to Lofi

  - Hex colours are converted to lofi representation by considering each
    of the three 8-bit colour channels and quantizing the value into
    fewer bits.
  - The number of bits can be specified individually for the separate R,
    G and B colours.
  - The folowing shows that the reconstructed colour isn’t identical to
    the original, but it is still a reasonable
approximation.

| Representation                                                              | Bits | Value    | Colour sample or bit layout      |
| --------------------------------------------------------------------------- | ---- | -------- | -------------------------------- |
| Original colour                                                             | 24   | \#123456 | ![](man/figures/swatch1.png)     |
| Low-fidelity colour `hex_colour_to_lofi('#123456', rgb_bits = c(3, 3, 2)))` | 8    | 5L       | ![](man/figures/colour-bits.png) |
| Reconstructed colour `lofi_to_hex_colour(5L,  rgb_bits = c(3, 3, 2))`       | 24   | \#002455 | ![](man/figures/swatch2.png)     |

For more example of using `lofi` to pack colours see
[`vignette("lofi-colour",
package='lofi')`](https://coolbutuseless.github.io/package/lofi/articles/lofi-colour.html)

## Integer to Lofi

  - Integers are converted to low-fidelity representation by truncacting
    any leading zeros (or, in the case of negative numbers, truncating
    the leading ones)
  - Keeping a leading bit for the sign is optional, and if it is
    excluded then the lofi representation is only able to hold positive
    numbers
  - `lofi` correctly keeps the sign bit and twos-complement for negative
    values

| Representation                                                       | Bits | Value | bit layout                        |
| -------------------------------------------------------------------- | ---- | ----- | --------------------------------- |
| Original integer                                                     | 32   | \-12  |                                   |
| Low-fidelity integer `int32_to_lofi(-12L, nbits = 5, signed = TRUE)` | 5    | 20L   | ![](man/figures/integer-bits.png) |
| Reconstructed integer `lofi_to_int32(20, nbits = 5, signed = TRUE)`  | 32   | \-12  |                                   |

## Choice to Lofi

  - Convert a value into a zero-based index into a list of options.
  - This is similar in idea to how a `factor` works in R

<!-- end list -->

``` r
options <- c('apple', 'banana', 'carrot', 'dog')
choice  <- c('apple', 'apple', 'dog')

(lofi   <- choice_to_lofi(choice, options))
#> [1] 0 0 3

lofi_to_choice(lofi  , options)
#> [1] "apple" "apple" "dog"
```

# Caveats + Future

  - The `pack/unpack` spec is limited to 32-bits as
    1.  this is my use case.
    2.  that’s the maximum number of bits in an R integer
    3.  `double` values are 64-bits but they don’t have built-in support
        for bitwise logical operations like `bitwAnd()` etc.
  - It should be possible to instead pack/unpack values into a unlimited
    stream of `raw` bytes. This may be considered for future versions.
  - Factors can not be directly packed. You will have to convert to
    another representation (e.g. character) and encode with
    `choice_to_lofi()`.


NEWS.md
# lofi 0.1.1

* Added some helper functions for colour/in32 conversion:
    * `rgba_vec_to_hex_colour()`
    * `hex_colour_to_rgba_vec()`
    * `int32_to_hex_colour()`
    * `hex_colour_to_int32()`
    * `rgba_vec_to_int32()`
    * `int32_to_rgba_vec()`

# lofi 0.1.0

* Initial release.


R/pack-spec-sanitise.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Expect the given value to be a single, non-NA numeric value
#'
#' @param value value
#' @param value_desc description of value to be inserted into error message
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
expect_single_numeric <- function(value, value_desc = "value") {
  if (length(value) != 1   ||
      !is.numeric(value)   ||
      is.na(value)) {
    stop("pack_spec: Invalid ", value_desc, ". Expecting single, non-NA numeric value. ",
         "Current value: ", deparse(value), call. = FALSE)
  }
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Validate 'nbits' to be a proper value
#'
#' @param nbits number of bits
#' @param value_desc description of value to be inserted into error message
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
expect_valid_nbits <- function(nbits, value_desc = 'nbits') {
  expect_single_numeric(nbits, value_desc)

  if (nbits < 1 || nbits > 32) {
    stop("pack_spec: Invalid ", value_desc, ".  Expecting nbits in range [1, 32]. ",
         "Current value: ", deparse(nbits), call. = FALSE)
  }
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for a 'integer'
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_integer <- function(spec) {
  stopifnot(identical(spec$type, 'integer'))

  spec$offset <- spec$offset %||% 0L
  spec$mult   <- spec$mult   %||% 1L
  spec$signed <- spec$signed %||% FALSE

  expect_single_numeric(spec$offset, "integer 'offset'")
  expect_single_numeric(spec$mult  , "integer 'mult'"  )
  expect_valid_nbits   (spec$nbits , "integer 'nbits'" )

  spec$nbits  <- round(spec$nbits)

  spec
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for a 'scaled' value
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_scaled <- function(spec) {
  stopifnot(identical(spec$type, 'scaled'))

  spec$min      <- spec$min      %||% 0L
  spec$cyclical <- spec$cyclical %||% FALSE

  if (is.null(spec$max)) {
    stop("sanitise_spec_scaled(): Invalid 'spec$max'.  Expecting number. ",
         "Current value: ", deparse(spec$max), call. = FALSE)
  }

  expect_single_numeric(spec$min   , "scaled 'min'"  )
  expect_single_numeric(spec$max   , "scaled 'max'"  )
  expect_valid_nbits   (spec$nbits , "scaled 'nbits'")

  spec$nbits  <- round(spec$nbits)

  spec
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for a 'custom' pack
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_custom <- function(spec) {
  stopifnot(identical(spec$type, 'custom'))

  spec$signed <- spec$signed %||% FALSE

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Convert to functions if supplied as formulas
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (inherits(spec$pack_func, 'formula')) {
    spec$pack_func <- formula_to_1arg_function(spec$pack_func)
  }

  if (inherits(spec$unpack_func, 'formula')) {
    spec$unpack_func <- formula_to_1arg_function(spec$unpack_func)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Check the pack_func and unpack_func are functions
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (is.null(spec$pack_func) || !is.function(spec$pack_func)) {
    stop("sanitise_spec_custom(): Invalid spec$pack_func. Expecting formula or function. ",
         "Current value: ", deparse(spec$pack_func), call. = FALSE)
  }

  if (is.null(spec$unpack_func) || !is.function(spec$unpack_func)) {
    stop("sanitise_spec_custom(): Invalid spec$unpack_func Expecting formula or function. ",
         "Current value: ", deparse(spec$unpack_func), call. = FALSE)
  }

  expect_valid_nbits(spec$nbits , "integer 'nbits'" )

  spec
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for a 'logical'
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_logical <- function(spec) {
  stopifnot(identical(spec$type, 'logical'))

  spec$nbits  <- spec$nbits %||% 1

  expect_valid_nbits(spec$nbits, "logical 'nbits'" )
  spec$nbits  <- round(spec$nbits)

  spec
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for a 'double'
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_double <- function(spec) {
  stopifnot(identical(spec$type, 'double'))

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # If user has specified an 'nbits' do an auto assigment based upon their
  # proposed maxval.
  #
  # If user has specified a 'float_name', then use that to define `float_bits`
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (!is.null(spec$nbits) && is.null(spec$float_bits)) {
    if (is.null(spec$maxval)) {
      stop("sanitise_spec_double(): must specify a 'maxval' if specifying 'nbits'", call. = FALSE)
    }
    spec$signed     <- spec$signed %||% FALSE
    spec$float_bits <- allocate_float_bits(spec$nbits, max_value = spec$maxval, signed = spec$signed)
    # message("Auto allocate double bits: ", spec$nbits, ": ", deparse(spec$float_bits))
  } else if (!is.null(spec$float_name) && spec$float_name %in% names(float_formats)) {
    spec$float_bits <- float_formats[[spec$float_name]]$float_bits
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Check that we have float_bits assigned to smething now.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (length(spec$float_bits) == 0 || !is.numeric(spec$float_bits)) {
    stop("sanitise_spec_double(): Invalid 'float_bits' for double. ",
         "Expecting non-NA numeric vector of length 3. Current value: ",
         deparse(spec$float_bits), call. = FALSE)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Ensure 'float_bits' is integer
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  spec$float_bits <- as.integer(spec$float_bits)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Make sure float_bits is sane
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (!length(spec$float_bits) == 3 || any(is.na(spec$float_bits)) ||
      any(spec$float_bits < 0) ||
      sum(spec$float_bits) > 32) {
    stop("sanitise_spec_double(): Invalid 'float_bits' for double. ",
         "Expecting non-NA numeric vector of length 3 with sum(float_bits) in range [0, 32]. ",
         "Current value: ", deparse(spec$float_bits), call. = FALSE)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Sum the bits for the float spec
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  spec$nbits  <- sum(spec$float_bits)
  expect_valid_nbits(spec$nbits, "double 'nbits'" )


  spec
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for an 'choice'
#'
#' @param spec named list object with 'type', 'nbits', etc
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_choice <- function(spec) {
  stopifnot(identical(spec$type, 'choice'))

  if (length(spec$options) == 0L) {
    stop("pack_spec: Invalid choice 'options'. Expecting at least one element. ",
         "Current value: ", call. = FALSE)
  }


  spec$nbits <- spec$nbits %||% ceiling(log2(length(spec$options)))
  if (spec$nbits == 0) {
    spec$nbits <- 1L
  }
  spec$nbits  <- round(spec$nbits)


  if (spec$nbits < ceiling(log2(length(spec$options)))) {
    stop("Invalid pack_spec: Not enough bits to hold all options. Bits: ",
         spec$nbits, "  Values: ", deparse(spec$options), call. = FALSE)
  }

  expect_valid_nbits(spec$nbits , "choice 'nbits'" )

  spec
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Usually the ser can just specify number of bits, and we'll pick a
# pretty good bit allocation for RGB.
# Can't do bits < 3, and can't do bits > 24
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
default_rgb_bits <- list(
  NA,
  NA,
  c(1, 1, 1),
  c(1, 2, 1),
  c(2, 2, 1),
  c(2, 2, 2),
  c(2, 3, 2),
  c(3, 3, 2),
  c(3, 3, 3),
  c(3, 4, 3),
  c(4, 4, 3),
  c(4, 4, 4),
  c(4, 5, 4),
  c(5, 5, 4),
  c(5, 5, 5),
  c(5, 6, 5),
  c(6, 6, 5),
  c(6, 6, 6),
  c(6, 7, 6),
  c(7, 7, 6),
  c(7, 7, 7),
  c(7, 8, 7),
  c(8, 8, 7),
  c(8, 8, 8),
  NA, NA, NA, NA, NA, NA, NA, NA
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a spec for an 'colour'
#'
#' @param spec named list object with 'type = "colour"', 'nbits' or 'rgb_bits'
#'
#' @return Return sanitised spec with any generic default values set
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec_colour <- function(spec) {
  stopifnot(identical(spec$type, 'colour'))

  if (!is.null(spec$rgb_bits)) {
    # use rgb bits as given
  } else if (is.null(spec$nbits) || length(spec$nbits) != 1L ||
             is.na(spec$nbits) || !is.numeric(spec$nbits)) {
    stop("sanitise_spec_colour(): Invalid 'nbits'. ",
         "Expecing 'nbits' to be single non-NA numeric value. ",
         "Current value: ", deparse(spec$nbits), call. = FALSE)
  } else if (spec$nbits < 3 || spec$nbits > 24) {
    stop("sanitise_spec_colour(): Invalid 'nbits'. ",
         "Expecting 'nbits' in range [3, 24]. ",
         "Current value: ", deparse(spec$nbits), call. = FALSE)
  } else {
    spec$rgb_bits <- default_rgb_bits[[spec$nbits]]
  }

  spec$nbits <- sum(spec$rgb_bits)
  expect_valid_nbits(spec$nbits , "colour 'nbits'")

  spec
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise an individual spec
#'
#' @param spec single spec
#'
#' @return Return sanitised spec
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_spec <- function(spec) {

  if (is.null(spec$type)) {
    warning("sanitise_spec(): spec contains no 'type'")
    print(spec)
    stop("sanitise_spec(): spec contains no 'type'")
  }

  switch(
    spec$type,
    integer = sanitise_spec_integer(spec),
    logical = sanitise_spec_logical(spec),
    double  = sanitise_spec_double (spec),
    choice  = sanitise_spec_choice (spec),
    colour  = sanitise_spec_colour (spec),
    custom  = sanitise_spec_custom (spec),
    scaled  = sanitise_spec_scaled (spec),
    stop("sanitise_spec(): Invalid spec$type: ", deparse(spec$type))
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise a full pack_spec
#'
#' @param pack_spec pack_spec
#'
#' @return Return sanitised pack_spec
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_pack_spec <- function(pack_spec) {
  pack_spec <- lapply(pack_spec, sanitise_spec)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Check bits
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  all_bits   <- vapply(pack_spec, function(x) {x$nbits}, FUN.VALUE = numeric(1))
  total_bits <- sum(all_bits)

  if (is.na(total_bits) || total_bits < 1 || total_bits > 32 || total_bits != round(total_bits)) {
    stop("validate_pack_spec(): Invalid total 'bits' in pack_spec. ",
         "Expecting total bits in range [1, 32]. ",
         "Current value: ", total_bits, call. = FALSE)
  }

  pack_spec
}

